### PR TITLE
Improved security groups in OpenAPI generation

### PR DESCRIPTION
**Pull request by inklesspen**
Audiences are moved from x-security to an x-google-audiences key in the security definition. We autogenerate extra security definitions as needed.

### CODE REVIEW

File: endpoints/openapi_generator.py [hunk 1]
```diff
@@ -839,7 +823,8 @@ def __api_openapi_descriptor(self, services, hostname=None):
     for service in services:
       remote_methods = service.all_remote_methods()
 
-      for protorpc_meth_name, protorpc_meth_info in remote_methods.iteritems():
+      for protorpc_meth_name in sorted(remote_methods.iterkeys()):
```
Comment (tangiel) [inline endpoints/openapi_generator.py]: What is the purpose of this change?

File: endpoints/test/openapi_generator_test.py [hunk 2]
```diff
@@ -1360,8 +1468,203 @@ def noop_get(self, unused_request):
...
+                        "type": "string"
+                    },
+                    "string_value": {
+                        "type": "string"
+                    },
+                },
+            },
+        },
+        "securityDefinitions": {
+            "auth0": {
```
Comment (tangiel) [inline endpoints/test/openapi_generator_test.py]: Is this definition necessary?

File: endpoints/openapi_generator.py (hunk 1 above)
Comment (inklesspen) [reply endpoints/openapi_generator.py]: The order of iteration over Python dictionaries is undefined; I wanted to have a stable ordering.

File: endpoints/test/openapi_generator_test.py (hunk 2 above)
Comment (inklesspen) [reply endpoints/test/openapi_generator_test.py]: No, but it does no harm.

The openapi generator directly mutates the ending dictionary structure instead of using a Builder pattern like the Java framework does. This means we need the un-modified security definition in the structure until we're done processing methods.

At that point it would certainly be possible to remove unused security definitions, but I felt that was an unnecessary optimization. If you disagree, I can add it.

**Comment by inklesspen** [comment]
tangiel approved merging via chat
